### PR TITLE
Await when token is refreshing & reinstate `msRefreshBeforeExpires` flag

### LINKED ISSUE #12648: SDK: Token refresh failed with concurrent requests
### Describe the Bug

Scenario (see also #12636):

1. A page of my vuejs webapp is open until the access token expires.
2. Then a router link is clicked. Multiple simultaneous api requests are executed.
3. The refresh request is triggered, but other requests fail with 403, because authorization header is empty.

The SDK is not waiting while a refresh is running.

I have a possible solution and will link a pull request.

### To Reproduce

Perform multiple requests with the SDK with expired access token.

### Errors Shown

_No response_

### What version of Directus are you using?

9.8.0

### What version of Node.js are you using?

16.14.2

### What database are you using?

Postgres 14

### What browser are you using?

Chrome, Edge, Firefox

### How are you deploying Directus?

Docker

### PR DESCRIPTION
**Pull request by lubrst**
Fix #12648

Refer discussion #12636. 

## Changes

- A reference to the refresh promise is used to wait in `refreshIfExpired()`.
- Re-added `msRefreshBeforeExpires` option.
- Removed nested refresh calls in `refresh()` and `login()`.
- Do not return `refresh_token` property when it does not exist to prevent potential confusion. Mentioned in https://github.com/directus/directus/issues/12427#issuecomment-1086604045.

### CODE REVIEW

**Review by azrikahar — CHANGES_REQUESTED** [review]
Out of curiosity, were you able to test this change locally with your own use case? 

I did a rudimentary test but it seems like we might also need to remove the `if (this.autoRefresh) this.refreshIfExpired();` lines from refresh() and login() to prevent an infinite loop when we are setting an ACCESS_TOKEN_TTL that is lower than 30s. This situation is an edge case in and of itself, but it was also one of the issue in the previous timer-based implementation.

I believe they were there for "resetting" the timer, but I think they are not needed anymore since the timer no longer exists, and refreshing/logging in should already yield new tokens in the first place. Just mentioning this here in case my test itself is missing something. cc @krazyjakee to get a second opinion.

**Comment by krazyjakee** [comment]
> infinite loop when we are setting an ACCESS_TOKEN_TTL that is lower than 30s
Technically that issue is being re-introduced with this PR so I would consider it in-scope.

I think the fix only needs to focus on the `refresh` function. Perhaps a simple check to ensure that `ACCESS_TOKEN_TTL` is set to `msRefreshBeforeExpires` if it is lower than `msRefreshBeforeExpires` if that makes sense?

**Comment by azrikahar** [comment]
> Technically that issue is being re-introduced with this PR so I would consider it in-scope.

Agreed!

> Perhaps a simple check to ensure that `ACCESS_TOKEN_TTL` is set to `msRefreshBeforeExpires` if it is lower than `msRefreshBeforeExpires` if that makes sense? Should also log out a warning

Yeah thought about this, and it was the approach taken via `Math.min()` in the other PR #12281 (before your PR #12399 landed), but that would still only "partially" works right? Understandably also probably why you mentioned to log a warning. A minimum does make sense, but not sure if that should be restricted upstream in the API side instead of SDK, and what would be the percentage of use cases with less than 30 or even 10 seconds access token TTL to begin with. 🤔

---

Just to confirm, what is your verdict on the removal of this two lines? Just want to be extra sure to prevent their removal from causing any issue with unforeseen edge cases that I am not be aware of here.

- inside `refresh()`:

   https://github.com/directus/directus/blob/985605b3564f8d9c0943dcd95ab7e08663c52435/packages/sdk/src/base/auth.ts#L88

- inside `login()`:

   https://github.com/directus/directus/blob/985605b3564f8d9c0943dcd95ab7e08663c52435/packages/sdk/src/base/auth.ts#L108

**Comment by lubrst** [comment]
> Out of curiosity, were you able to test this change locally with your own use case?

Yes, in my setup (own vue.js web app) the changes work.

> I did a rudimentary test but it seems like we might also need to remove the if (this.autoRefresh) this.refreshIfExpired(); lines from refresh() and login() to prevent an infinite loop when we are setting an ACCESS_TOKEN_TTL that is lower than 30s.

I also can reproduce the infinit loop behavior with ACCESS_TOKEN_TTL lower than 30s. If I remove `if (this.autoRefresh) this.refreshIfExpired();` lines from refresh() and login() no infinit loop occurs and before every request a refresh is performed (the bevhavior I would expect).

**Comment by lubrst** [comment]
The failed unit test in auth.node.test.ts could be resolved by removing the `if (this.autoRefresh) this.refreshIfExpired(); ` lines and adding missing `expires: 60000` to the `/auth/login` mock data (or adding a `/auth/refresh` mock).

**Comment by azrikahar** [comment]
@lubrst thanks for the confirmation! As well as the recommend resolution for the unit test. I'll make the necessary changes 👍

**Review by azrikahar — APPROVED** [review]

**Review by rijkvanzanten — APPROVED** [review]

**Pull request merged**
